stun.go

```diff
@@ -67,35 +103,45 @@ func assertInboundMessageIntegrity(m *stun.Message, key []byte) error {
 	return nil
 }
 
-func allocateUDP(network string, url *URL) (*net.UDPAddr, *stun.XorAddress, error) {
-	// TODO Do we want the timeout to be configurable?
-	client, err := stun.NewClient(network, fmt.Sprintf("%s:%d", url.Host, url.Port), time.Second*5)
+func getServerReflexiveAddress(conn *net.UDPConn, network string, url *URL, deadline time.Duration) (*stun.XorAddress, error) {
+	hostPort := fmt.Sprintf("%s:%d", url.Host, url.Port)
+	stunAddr, err := net.ResolveUDPAddr(network, hostPort)
 	if err != nil {
-		return nil, nil, flattenErrs([]error{errors.New("failed to create STUN client"), err})
-	}
-	localAddr, ok := client.LocalAddr().(*net.UDPAddr)
-	if !ok {
-		return nil, nil, fmt.Errorf("failed to cast STUN client to UDPAddr")
+		return nil, fmt.Errorf("failed to resolve stun host: %s: %v", hostPort, err)
 	}
 
-	resp, err := client.Request()
+	err = conn.SetReadDeadline(time.Now().Add(deadline))
+	if err != nil {
+		return nil, err
+	}
+	err = conn.SetWriteDeadline(time.Now().Add(deadline))
 	if err != nil {
-		return nil, nil, flattenErrs([]error{errors.New("failed to make STUN request"), err})
+		return nil, err
 	}
 
-	if err = client.Close(); err != nil {
-		return nil, nil, flattenErrs([]error{errors.New("failed to close STUN client"), err})
+	resp, err := stun.Request(&stunConn{conn, stunAddr})
```

Comment:
Maybe instead of stun.Request i could add a GetMappedAddress method that does everything that request does, but also reads and returns the XorAddress. What do you think?

Reply:
Yea 👍 from me!

We should try and keep `pion/ice` as small as possible, it is about to get much more bloated with TURN coming :/